fix(razzle): Improve SIGINT/SIGTERM handling.

therealgilles opened the pull request:
start.js:
- Close server watcher on SIGINT/SIGTERM.

StartServerPlugin.js:
- On SIGINT, if the child process SIGINT handler is not in place yet or 5 seconds have elapsed, exit right away.
- On non-SIGTERM signals, if the child process SIGINT handler is in place, exit once the child process has exited. This allows the (http) server to go through its SIGINT shutdown sequence if any.

therealgilles committed:
fix(razzle): Improve SIGINT/SIGTERM handling.

razzle/package.json:
- Add terminate package as dependency.

start.js:
- Close server watcher on SIGINT/SIGTERM.
- Use terminate package to handle SIGINT by sending SIGINT to any child
  process with a 1-second timeout, then terminate the current process.
  Turns out process.kill(process.pid) is better then process.exit() to
  avoid messing up with the terminal on exit.

StartServerPlugin.js:
- Do not use process.exit() once child process is done. Exit is now
  handled by the terminate package in start.js.
- Identify SIGTERM through signal name or code 143.

therealgilles commented:
@fivethreeo: I was not able to get rid of the global var in StartServerPlugin.js. The reason I use it is that, if the child process handler is not set, we need to exit right away. But if the child process handler is set, we exit only when the child process has exited, or after 5 seconds if somehow the child process is stuck or is taking too long to exit.

I don't think it should be an issue but let me know why it could be a problem and/or if you have a better idea :)

fivethreeo commented:
The issue is if someone wants to use two plugins. Could we use a set and make each plugin instance have a UUID?

therealgilles commented:
Do you mean we may have two instances of StartServerPlugin? or in a more general case, we may have multiple child processes running?

I think ideally, we would use a tool like ps-tree in start.js, to properly handle all child processes at once or exit right away if none are found. I can work on that. What do you think?

therealgilles commented:
Another option would be to use the [terminate package](https://github.com/dwyl/terminate) (which relies on ps-tree). I will run some experiments. I don't know if you would be okay with the added dependencies.

fivethreeo commented:
More stable is better, deps are ok :)

therealgilles commented:
What's the right flow to add a dependency to packages/razzle/package.json? Should I do a 'yarn add terminate' in that directory and then 'yarn updateyalc' in the example dir?

fivethreeo commented:
npm add lerna-update-wizard -g
then lernaupdate in the razzle monorepo
then 'yarn updateyalc' in the example dir
then 'NODE_ENV=development yarn install ---ignore-engines' in the example dir

On Sat, 23 Jan 2021 at 21:09, Gilles <notifications@github.com> wrote:

> What's the right flow to add a dependency to packages/razzle/package.json?
> Should I do a 'yarn add terminate' in that directory and then 'yarn
> updateyalc' in the example dir?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jaredpalmer/razzle/pull/1528#issuecomment-766172067>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAADGRGLPMLPLD2KSAOZM3S3MUITANCNFSM4WPPDAGA>
> .
>


-- 
Øyvind Saltvik

therealgilles commented:
Is it expected that my changes affect both the src and dist version of StartServerPlugin.js? I think so, but I want to confirm in case I did something wrong (it was not the case in my original pull request but maybe I had missed something there):

```JS
# On branch feature/signal-handlers
# Changes to be committed:
#       modified:   packages/razzle-start-server-webpack-plugin/dist/StartServerPlugin.js
#       modified:   packages/razzle-start-server-webpack-plugin/src/StartServerPlugin.js
#       modified:   packages/razzle/package.json
#       modified:   packages/razzle/scripts/start.js
#       modified:   yarn.lock
```

fivethreeo commented:
correct

therealgilles commented:
Submitted the updated PR. Here is the log:

```GIT
razzle/package.json:
- Add terminate package as dependency.

start.js:
- Close server watcher on SIGINT/SIGTERM.
- Use terminate package to handle SIGINT by sending SIGINT to any child process with a 1-second timeout,
  then terminate the current process. Turns out process.kill(process.pid) is better then process.exit()
  to avoid messing up with the terminal on exit.

StartServerPlugin.js:
- Do not use process.exit() once child process is done. Exit is now handled by the terminate package
  in start.js.
- Identify SIGTERM through signal name or code 143.
```

fivethreeo commented:
I would like this to be useable outside razzle aswell. This breaks that. Maybe we could use sone webpack hook in the plugin to force exit at the right time?

therealgilles commented:
Do you mean you would like razzle-start-server-plugin to be usable outside of razzle?

fivethreeo commented:
Yes, forked it since maintainers were mia. It should be just a webpack plugin. We could make it have a option to make it work both ways. Just not as nicely until a better way is found. If possible.

therealgilles commented:
You mean you would like to keep the process.exit() by default -- or maybe replace it with process.kill(process.id). And then have an option to not let it exit?

From a process management viewpoint, the hierarchy will always look something like this:

     main process / node entry point [start.js for razzle]
        |
        |--- plugin #1
        |          |----  forked child process
        |
        |--- plugin #2
        |          |----  forked child process
        ....

It's a bit odd that a plugin would be able to issue a process.exit() when it's not the entry point of the main process. If several plugins are doing this, one of them will win the race and the other ones will be interrupted in their exit sequences.

therealgilles commented:
I saw you added process.exit() with this checkin:

```
commit 2d6a9bd195ea745180b42f2464124c55f98d366e
Author: Øyvind Saltvik
Date:   Fri Jan 22 19:05:59 2021 +0100

     fix(razzle-start-server-webpack-plugin): fix error after process exit
```

What were you trying to fix here? If it was because the process was not exiting, that's because watching.close() had not been called, which I added in the PR.

If you want to add an option, this would be flexible though a little ugly:
```JS
    this.options = Object.assign(
      {
        ...
        onChildProcessExit: null, // () => { process.exit(); } or () => { process.kill(process.pid, 'SIGKILL'); }
      },
      options
    );
...
  _handleChildExit(code, signal) {
    ...
    if (code === 143 || signal === 'SIGTERM') {
      ...
      this._runWorker();
      return;
    }

    if (typeof this.options.onChildProcessExit === 'function') {
      this.options.onChildProcessExit();
    }
  }
```

fivethreeo commented on packages/razzle-start-server-webpack-plugin/src/StartServerPlugin.js:
```diff
@@ -118,26 +118,22 @@ export default class StartServerPlugin {
   }
 
   _handleChildExit(code, signal) {
```
https://nodejs.org/api/child_process.html#child_process_event_error

fivethreeo commented on packages/razzle-start-server-webpack-plugin/src/StartServerPlugin.js:
```diff
@@ -118,26 +118,22 @@ export default class StartServerPlugin {
-      process.exit()
-      return;
-    }
-    if (!this.workerLoaded) {
```
Handle this in error handler aswell?

therealgilles replied:
Do you mean provide a way to also call process.exit() in the child process error handler instead of just setting this.worker to null? (and then safeguard for the case where both exit and error are triggered)

fivethreeo commented:
> I saw you added process.exit() with this checkin:
> 
> ```
> commit 2d6a9bd195ea745180b42f2464124c55f98d366e
> Author: Øyvind Saltvik
> Date:   Fri Jan 22 19:05:59 2021 +0100
> 
>      fix(razzle-start-server-webpack-plugin): fix error after process exit
> ```
> 
> What were you trying to fix here? If it was because the process was not exiting, that's because watching.close() had not been called, which I added in the PR.
> 
> ```

Was a attempt at not getting the tty error.

therealgilles commented:
Oh I see. Found this in nodejs documentation:

[https://nodejs.org/api/process.html#process_signal_events](https://nodejs.org/api/process.html#process_signal_events)
'SIGTERM' and 'SIGINT' have default handlers on non-Windows platforms that reset the terminal mode before exiting with code 128 + signal number. If one of these signals has a listener installed, its default behavior will be removed (Node.js will no longer exit).

From what I understand now, this is not the full truth (see nodejs open issues above). It seems the only way to avoid having nodejs mess with the terminal on exit is by using SIGKILL for the termination (which is what I did with terminate in the PR).

fivethreeo commented:
```js

  _handleChildError(err) {
    this.worker = null;
    this.workerLoaded = false;
    this._error('Script errored');
    if (this.options.terminateonerror) {
     terminate(process.pid);
    }
  }

```

therealgilles committed:
fix(razzle): Improve SIGINT/SIGTERM handling.

StartServerPlugin.js:
- Add killOnExit, killOnError, and killTimeout options.
  If enabled, sending SIGKILL to process after killTimeout or
  immediately if null. Default to true/true/1000 milliseconds.

createConfigAsync.js:
- Set killOnExit and killOnError to false as we are handling the process
  termination locally with terminate.

therealgilles commented:
Here is my update:

```
    StartServerPlugin.js:
    - Add killOnExit, killOnError, and killTimeout options.
      If enabled, sending SIGKILL to process after killTimeout or
      immediately if null. Default to true/true/1000 milliseconds.
    
    createConfigAsync.js:
    - Set killOnExit and killOnError to false as we are handling the process
      termination locally with terminate.
```

fivethreeo commented:
Think we are getting somewhere here :)

fivethreeo commented:
Hopefully here too

https://github.com/jaredpalmer/razzle/milestone/1

fivethreeo commented:
May need startserver changes for server components.

therealgilles commented:
For webpack 5 module federation, setups will have more ports. Do you think port configuration in razzle will need revamping?
React server components seem a few months away from being ready to integrate into tools.

fivethreeo commented:
https://github.com/jaredpalmer/razzle/blob/canary/examples/with-module-federation/package.json#L11

fivethreeo commented:
React server components just needs to be able to  not add the monitor in
startserver or just a stub monitor with loaded message.

On Tue, 26 Jan 2021 at 08:39, Gilles <notifications@github.com> wrote:

> For webpack 5 module federation, setups will have more ports. Do you think
> port configuration in razzle will need revamping?
> React server components seem a few months away from being ready to
> integrate into tools.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jaredpalmer/razzle/pull/1528#issuecomment-767362395>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAADGVK72VYV5JBMS2HCPDS3ZWTPANCNFSM4WPPDAGA>
> .
>


-- 
Øyvind Saltvik

fivethreeo commented:
https://github.com/sibelius/react-server-components-playground
https://github.com/jaredpalmer/razzle/tree/canary/examples/with-react-server-components

therealgilles commented:
I've been thinking about nodejs signal handling some more:

`'SIGTERM' and 'SIGINT' have default handlers on non-Windows platforms that reset the terminal mode before exiting with code 128 + signal number. If one of these signals has a listener installed, its default behavior will be removed (Node.js will no longer exit).`

Maybe I got it reversed. That is without the SIGINT handler, node may reset the terminal mode properly on exit. But as we need the SIGINT handler for the server to run its own handler, we are disabling the default node handler and the terminal mode is not getting reset anymore. But I don't see any access to a 'reset terminal mode' function anywhere. It's quite odd that exiting is an issue for a tool that has been around for so long. Imagine if the terminal was messed up every time you ran a shell command.

fivethreeo commented:
I tried this, left over processes after that

```bash
pwd
# /home/oyvind/Documents/GitHub/razzle/examples/basic
example="$(basename $PWD)"
cd ..
cd ..
yarn bootstrap-examples $example
cd examples/$example
yarn build
timeout 10 yarn start -s INT &
fg %1
```

therealgilles commented:
Hmm I don't seem to have bootstrap-examples in package.json. Also don't you have to write: `timeout -s INT 10 yarn start` ?

PS: Also a bit confused about the `yarn build`. Isn't that for the production build? whereas `yarn start` will start the development build.

therealgilles commented:
If I run this:
```SHELL
> ps; gtimeout -s SIGINT 10 yarn start_yarn; sleep 2; ps
```

I get the following:
```SHELL
 2261 ttys000    0:00.45 -zsh
yarn run v1.22.10
$ razzle start
 WAIT  Compiling...


✔ Client
  Compiled successfully in 1.63s

✔ Server
  Compiled successfully in 285.75ms

✅  Server-side HMR Enabled!
sswp> Handling Hot Module Reloading
> Started on port 3000
 SIGINT  exiting...
Server received SIGINT
sswp> !!! Script exited with signal SIGINT
 Goodbye 
  PID TTY           TIME CMD
 2261 ttys000    0:00.45 -zsh
```

zsh is my shell.

When you ran your command, did you see the ` SIGINT  exiting...` and ` Goodbye ` messages?
Which processes are left behind for you?

fivethreeo commented:
Just do a rebase on canary

Just left that in, was trying to automate examples testing

fivethreeo commented:
Maybe I got the merge wrong? :)

therealgilles commented:
Rebase done and pull request updated. The `  SIGINT  exiting...` message should show in red and ' Goodbye ' in green.

<img width="306" alt="Screen Shot 2021-01-29 at 5 17 34 PM" src="https://user-images.githubusercontent.com/18076992/106342487-e5147f00-6255-11eb-9625-2eab99b0af28.png">

fivethreeo commented:
Merged, great work :)

therealgilles commented:
Thank you for your patience and guidance. Really appreciated.
